File: src/benchmark/dimensions/HierarchicalCoordination.js
// HierarchicalCoordination — cross-channel agreement dimension plugin (S6).
//
// Score families (mixed scales — DO NOT pipe directly into bayesianScore):
//   - plan_policy_alignment    ∈ [0,1]  higher is better; fraction of strategic-
//                                       plan tokens (priority / resourceFocus /
//                                       workerFocus / phase / defensePosture)
//                                       that appear with positive weight in the
//                                       workers group's directive (intentWeights ∪
//                                       targetPriorities ∪ focus). Wired W3.
//                                       Tracks whether the strategic-plan channel
//                                       and the npc-policy channel agree on
//                                       which axes to prioritize.
//   - env_threat_responsiveness ∈ [-1,1] higher is better; Pearson corr of
//                                        factionTension vs threat
//   - colony_cadence_health     ∈ [0,∞) std-dev of decision intervals (sec);
//                                        lower is better; needs scenario-relative
//                                        normalization before bayesianScore
//
// Tests whether the 4 LLM channels behave coherently.

/** Map strategic-plan field values to the directive-token vocabulary used by
 *  Guardrails default group policies (intent / target keys, lowercase). The
 *  match is one-direction: each plan field contributes a token; alignment is
 *  the fraction of plan tokens that show up in the workers directive. */
function planTokens(strategy) {
  if (!strategy || typeof strategy !== "object") return [];
  const tokens = [];
  // resourceFocus directly maps to canonical intent / target keys.
  const rf = String(strategy.resourceFocus ?? "").toLowerCase();
  if (rf === "food") tokens.push("farm");
  else if (rf === "wood") tokens.push("wood", "lumber");
  else if (rf === "stone") tokens.push("quarry", "stone");
  // workerFocus is already in the intent vocabulary.
  const wf = String(strategy.workerFocus ?? "").toLowerCase();
  if (wf && wf !== "balanced") tokens.push(wf);
  // priority maps to high-level objective: "defend"→safety, "survive"→eat/safety
  // "complete_objective"→deliver. "grow" has no direct directive token.
  const pri = String(strategy.priority ?? "").toLowerCase();
  if (pri === "defend") tokens.push("safety");
  else if (pri === "survive") tokens.push("eat", "safety");
  else if (pri === "complete_objective") tokens.push("deliver");
  // phase: industrialize→quarry/smith, process→cook, fortify→safety, optimize→deliver
  const phase = String(strategy.phase ?? "").toLowerCase();
  if (phase === "industrialize") tokens.push("quarry", "smith");
  else if (phase === "process") tokens.push("cook");
  else if (phase === "fortify") tokens.push("safety");
  else if (phase === "optimize") tokens.push("deliver");
  // defensePosture: aggressive/defensive both raise safety as a directive token.
  const dp = String(strategy.defensePosture ?? "").toLowerCase();
  if (dp === "defensive" || dp === "aggressive") tokens.push("safety");
  // Dedup, keep order.
  return Array.from(new Set(tokens.filter(Boolean)));
}

/** Extract the directive token set with positive weight for a single group's
 *  policy entry (`{ intentWeights, targetPriorities, focus, … }`). */
function directiveTokenSet(policy) {
  const out = new Set();
  if (!policy || typeof policy !== "object") return out;
  for (const [k, w] of Object.entries(policy.intentWeights ?? {})) {
    if (Number(w) > 0) out.add(String(k).toLowerCase());
  }
  for (const [k, w] of Object.entries(policy.targetPriorities ?? {})) {
    if (Number(w) > 0) out.add(String(k).toLowerCase());
  }
  // focus is a free-text field, but tokens like "depot"/"tools" still register
  // when split on whitespace.
  for (const tok of String(policy.focus ?? "").toLowerCase().split(/\s+/)) {
    if (tok) out.add(tok);
  }
  return out;
}

function pearson(xs, ys) {
  const n = Math.min(xs.length, ys.length);
  if (n < 2) return 0;
  const mx = xs.slice(0, n).reduce((a, b) => a + b, 0) / n;
  const my = ys.slice(0, n).reduce((a, b) => a + b, 0) / n;
  let num = 0, dx2 = 0, dy2 = 0;
  for (let i = 0; i < n; i++) {
    const dx = xs[i] - mx;
    const dy = ys[i] - my;
    num += dx * dy;
    dx2 += dx * dx;
    dy2 += dy * dy;
  }
  if (dx2 === 0 || dy2 === 0) return 0;
  return num / Math.sqrt(dx2 * dy2);
}

export const HierarchicalCoordinationPlugin = {
  id: "hierarchical",
  label: "Hierarchical Coordination",
  scoreDimensions: [
    "plan_policy_alignment",
    "env_threat_responsiveness",
    "colony_cadence_health",
  ],

  async collectSamples(harness, opts = {}) {
    const durationSec = Number(opts.durationSec ?? 600);
    const samples = [];
    const totalTicks = Math.max(1, Math.round(durationSec / (1 / 30)));
    const sampleEveryTicks = Math.max(1, Math.round(10 / (1 / 30)));

    let lastColonyDecisionAt = 0;
    const colonyIntervals = [];

    for (let t = 0; t < totalTicks; t++) {
      await harness.tick();
      if (t % sampleEveryTicks !== 0) continue;
      const s = harness.state;
      const colonyDecisionAt = s.ai?.colonyAgent?.lastDecisionSec ?? 0;
      if (colonyDecisionAt > lastColonyDecisionAt) {
        colonyIntervals.push(colonyDecisionAt - lastColonyDecisionAt);
        lastColonyDecisionAt = colonyDecisionAt;
      }

      // Snapshot strategy + workers directive for plan_policy_alignment.
      const strategy = s.ai?.strategy ?? null;
      const policiesRaw = s.ai?.groupPolicies;
      let workerPolicy = null;
      if (policiesRaw instanceof Map) {
        const wrap = policiesRaw.get("workers");
        workerPolicy = wrap?.data ?? wrap ?? null;
      } else if (policiesRaw && typeof policiesRaw === "object") {
        const wrap = policiesRaw.workers;
        workerPolicy = wrap?.data ?? wrap ?? null;
      }

      samples.push({
        t: s.metrics.timeSec ?? 0,
        factionTension: s.ai?.environmentDirective?.factionTension ?? 0,
        threat: s.gameplay?.threat ?? 0,
        prosperity: s.gameplay?.prosperity ?? 0,
        // Snapshot for plan_policy_alignment — only the keys the alignment
        // computation reads; cheap and JSON-stable.
        strategySnapshot: strategy ? {
          priority: strategy.priority,
          resourceFocus: strategy.resourceFocus,
          workerFocus: strategy.workerFocus,
          phase: strategy.phase,
          defensePosture: strategy.defensePosture,
        } : null,
        workerPolicySnapshot: workerPolicy ? {
          intentWeights: { ...(workerPolicy.intentWeights ?? {}) },
          targetPriorities: { ...(workerPolicy.targetPriorities ?? {}) },
          focus: workerPolicy.focus ?? "",
        } : null,
      });
      if (s.session?.phase === "end") break;
    }
    samples._colonyIntervals = colonyIntervals;
    return samples;
  },

  selfScore(samples) {
    if (!samples?.length) {
      return {
        plan_policy_alignment: 0,
        env_threat_responsiveness: 0,
        colony_cadence_health: 0,
      };
    }
    const factionSeries = samples.map(s => s.factionTension);
    const threatSeries = samples.map(s => s.threat);
    const corr = pearson(factionSeries, threatSeries);

    const intervals = samples._colonyIntervals ?? [];
    let stdDev = 0;
    if (intervals.length > 1) {
      const m = intervals.reduce((a, b) => a + b, 0) / intervals.length;
      const variance = intervals.reduce((a, b) => a + (b - m) ** 2, 0) / intervals.length;
      stdDev = Math.sqrt(variance);
    }

    // ── plan_policy_alignment: per-sample fraction of plan tokens that
    // appear with positive weight in the workers directive. We average
    // across samples that have BOTH a strategy snapshot and a workers
    // directive (skipping samples where either is null — early ticks
    // before either channel has fired). If no sample has both, alignment
    // is 0 (= no observed agreement, never NaN).
    let alignSum = 0, alignCount = 0;
    for (const s of samples) {
      const strategy = s.strategySnapshot;
      const policy = s.workerPolicySnapshot;
      if (!strategy || !policy) continue;
      const tokens = planTokens(strategy);
      if (tokens.length === 0) continue; // strategy provides no testable tokens
      const directiveTokens = directiveTokenSet(policy);
      let hits = 0;
      for (const tok of tokens) if (directiveTokens.has(tok)) hits++;
      alignSum += hits / tokens.length;
      alignCount++;
    }
    const alignment = alignCount > 0 ? alignSum / alignCount : 0;

    return {
      plan_policy_alignment: Number(alignment.toFixed(4)),
      env_threat_responsiveness: Number(corr.toFixed(4)),
      colony_cadence_health: Number(stdDev.toFixed(2)),
    };
  },
};

File: project_utopia/benchmark/dimensions/hierarchical_coordination.py
"""Hierarchical Coordination dimension plugin — Python port (S6 / RC3 G3).

Mirrors ``src/benchmark/dimensions/HierarchicalCoordination.js`` including
the RC3 G3 ``plan_policy_alignment`` wiring (preserved EXACTLY).

Score families (mixed scales — DO NOT pipe directly into bayesian_score):

- ``plan_policy_alignment`` ∈ [0, 1] — Per-sample fraction of strategic-
  plan tokens (derived from ``state.ai.strategy.{priority, resource_focus,
  worker_focus, phase, defense_posture}``) that appear with positive weight
  in the workers group's directive
  (``intent_weights`` ∪ ``target_priorities`` ∪ focus). Averaged across
  samples that have both snapshots.
- ``env_threat_responsiveness`` ∈ [-1, 1] — Pearson correlation of
  ``faction_tension`` vs ``threat``.
- ``colony_cadence_health`` ∈ [0, ∞) — std-dev of decision intervals
  (lower is better).
"""

from __future__ import annotations

import math
import re
from typing import Any

from ..framework.dimension_plugin import DimensionPlugin

TICKS_PER_SEC = 30.0

_WHITESPACE_RE = re.compile(r"\s+")


def _state_get(state: Any, key: str, default: Any = None) -> Any:
    if isinstance(state, dict):
        return state.get(key, default)
    return getattr(state, key, default)


def _coerce_dict_or_map(raw: Any) -> dict[str, Any]:
    if raw is None:
        return {}
    if isinstance(raw, dict):
        return raw
    try:
        return dict(raw.items())
    except (AttributeError, TypeError):
        return {}


def _plan_tokens(strategy: dict[str, Any] | None) -> list[str]:
    """Map strategic-plan field values to directive-token vocabulary.

    One-direction match: each plan field contributes one or more tokens;
    alignment is the fraction of plan tokens present in the workers
    directive. Preserves order and dedupes (matches JS port's
    ``Array.from(new Set(...))`` semantics).
    """
    if not strategy or not isinstance(strategy, dict):
        return []
    tokens: list[str] = []
    # resource_focus directly maps to canonical intent / target keys.
    rf = str(strategy.get("resource_focus", strategy.get("resourceFocus", "")) or "").lower()
    if rf == "food":
        tokens.append("farm")
    elif rf == "wood":
        tokens.extend(("wood", "lumber"))
    elif rf == "stone":
        tokens.extend(("quarry", "stone"))
    # worker_focus is already in the intent vocabulary.
    wf = str(strategy.get("worker_focus", strategy.get("workerFocus", "")) or "").lower()
    if wf and wf != "balanced":
        tokens.append(wf)
    # priority maps to high-level objective.
    pri = str(strategy.get("priority", "") or "").lower()
    if pri == "defend":
        tokens.append("safety")
    elif pri == "survive":
        tokens.extend(("eat", "safety"))
    elif pri == "complete_objective":
        tokens.append("deliver")
    # phase
    phase = str(strategy.get("phase", "") or "").lower()
    if phase == "industrialize":
        tokens.append("quarry")
    elif phase == "fortify":
        tokens.append("safety")
    elif phase == "optimize":
        tokens.append("deliver")
    # defense_posture
    dp = str(strategy.get("defense_posture", strategy.get("defensePosture", "")) or "").lower()
    if dp in ("defensive", "aggressive"):
        tokens.append("safety")
    # Dedup, keep order.
    seen: set[str] = set()
    out: list[str] = []
    for tok in tokens:
        if tok and tok not in seen:
            seen.add(tok)
            out.append(tok)
    return out


def _directive_token_set(policy: dict[str, Any] | None) -> set[str]:
    """Extract directive-token set with positive weight from one group policy."""
    out: set[str] = set()
    if not policy or not isinstance(policy, dict):
        return out
    iw = policy.get("intent_weights", policy.get("intentWeights")) or {}
    tp = policy.get("target_priorities", policy.get("targetPriorities")) or {}
    for k, w in iw.items():
        try:
            if float(w) > 0:
                out.add(str(k).lower())
        except (TypeError, ValueError):
            continue
    for k, w in tp.items():
        try:
            if float(w) > 0:
                out.add(str(k).lower())
        except (TypeError, ValueError):
            continue
    focus = str(policy.get("focus", "") or "").lower()
    for tok in _WHITESPACE_RE.split(focus):
        if tok:
            out.add(tok)
    return out


def _pearson(xs: list[float], ys: list[float]) -> float:
    n = min(len(xs), len(ys))
    if n < 2:
        return 0.0
    mx = sum(xs[:n]) / n
    my = sum(ys[:n]) / n
    num = 0.0
    dx2 = 0.0
    dy2 = 0.0
    for i in range(n):
        dx = xs[i] - mx
        dy = ys[i] - my
        num += dx * dy
        dx2 += dx * dx
        dy2 += dy * dy
    if dx2 == 0 or dy2 == 0:
        return 0.0
    return num / math.sqrt(dx2 * dy2)


class HierarchicalCoordinationPlugin(DimensionPlugin):
    """Hierarchical Coordination plugin — RC3 G3 ``plan_policy_alignment`` wired."""

    id: str = "hierarchical"
    label: str = "Hierarchical Coordination"
    score_dimensions: tuple[str, ...] = (
        "plan_policy_alignment",
        "env_threat_responsiveness",
        "colony_cadence_health",
    )

    async def collect_samples(
        self,
        harness: Any,
        opts: dict[str, Any] | None = None,
    ) -> list[dict[str, Any]]:
        opts = opts or {}
        duration_sec = float(opts.get("duration_sec", opts.get("durationSec", 600)))
        samples: list[dict[str, Any]] = []
        total_ticks = max(1, round(duration_sec * TICKS_PER_SEC))
        sample_every_ticks = max(1, round(10 * TICKS_PER_SEC))

        last_colony_decision_at = 0.0
        colony_intervals: list[float] = []

        for t in range(total_ticks):
            await harness.tick()
            if t % sample_every_ticks != 0:
                continue
            s = harness.state
            ai = _state_get(s, "ai", None) or {}
            colony_agent = (ai.get("colony_agent", ai.get("colonyAgent")) or {})
            colony_decision_at = float(colony_agent.get("last_decision_sec", colony_agent.get("lastDecisionSec", 0)) or 0)
            if colony_decision_at > last_colony_decision_at:
                colony_intervals.append(colony_decision_at - last_colony_decision_at)
                last_colony_decision_at = colony_decision_at

            strategy = ai.get("strategy") or None
            policies_raw = _coerce_dict_or_map(ai.get("group_policies", ai.get("groupPolicies")))
            worker_policy = None
            if "workers" in policies_raw:
                wrap = policies_raw["workers"]
                worker_policy = wrap.get("data") if isinstance(wrap, dict) and "data" in wrap else wrap

            metrics = _state_get(s, "metrics", None) or {}
            env = ai.get("environment_directive", ai.get("environmentDirective")) or {}
            gameplay = _state_get(s, "gameplay", None) or {}
            session = _state_get(s, "session", None) or {}

            strategy_snapshot: dict[str, Any] | None = None
            if strategy:
                strategy_snapshot = {
                    "priority": strategy.get("priority"),
                    "resource_focus": strategy.get("resource_focus", strategy.get("resourceFocus")),
                    "worker_focus": strategy.get("worker_focus", strategy.get("workerFocus")),
                    "phase": strategy.get("phase"),
                    "defense_posture": strategy.get("defense_posture", strategy.get("defensePosture")),
                }
            worker_policy_snapshot: dict[str, Any] | None = None
            if worker_policy and isinstance(worker_policy, dict):
                worker_policy_snapshot = {
                    "intent_weights": dict(worker_policy.get("intent_weights", worker_policy.get("intentWeights")) or {}),
                    "target_priorities": dict(worker_policy.get("target_priorities", worker_policy.get("targetPriorities")) or {}),
                    "focus": worker_policy.get("focus", "") or "",
                }

            samples.append({
                "t": float(metrics.get("time_sec", metrics.get("timeSec", 0)) or 0),
                "faction_tension": float(env.get("faction_tension", env.get("factionTension", 0)) or 0),
                "threat": float(gameplay.get("threat", 0) or 0),
                "prosperity": float(gameplay.get("prosperity", 0) or 0),
                "strategy_snapshot": strategy_snapshot,
                "worker_policy_snapshot": worker_policy_snapshot,
            })
            if session.get("phase") == "end":
                break

        # Attach intervals — Python lists allow attributes via wrapping; we
        # return a plain list and store the intervals on a sentinel key in
        # the last sample so callers can pull them out without surprises.
        # The JS port set `samples._colonyIntervals` on the Array; here we
        # use a dict envelope through ctx in self_score, or attach via a
        # dedicated key. Because samples is iterated by index downstream we
        # cannot stash on the list — instead append a hidden sentinel sample.
        # Simpler: store as a class attr on self for the duration of one
        # collect+score cycle (single-threaded plugin contract).
        self._last_colony_intervals = list(colony_intervals)
        return samples

    def self_score(
        self,
        samples: list[dict[str, Any]],
        ctx: dict[str, Any] | None = None,
    ) -> dict[str, float]:
        if not samples:
            return {
                "plan_policy_alignment": 0.0,
                "env_threat_responsiveness": 0.0,
                "colony_cadence_health": 0.0,
            }
        faction_series = [float(s.get("faction_tension", 0.0)) for s in samples]
        threat_series = [float(s.get("threat", 0.0)) for s in samples]
        corr = _pearson(faction_series, threat_series)

        intervals = getattr(self, "_last_colony_intervals", None)
        if ctx and "colony_intervals" in ctx:
            intervals = list(ctx["colony_intervals"])
        intervals = intervals or []
        std_dev = 0.0
        if len(intervals) > 1:
            m = sum(intervals) / len(intervals)
            variance = sum((x - m) ** 2 for x in intervals) / len(intervals)
            std_dev = math.sqrt(variance)

        # plan_policy_alignment: per-sample fraction of plan tokens that
        # appear with positive weight in the workers directive. Averaged
        # across samples that have BOTH a strategy snapshot AND a workers
        # directive (skipping samples where either is None). Returns 0 if
        # no sample has both — never NaN.
        align_sum = 0.0
        align_count = 0
        for s in samples:
            strategy = s.get("strategy_snapshot")
            policy = s.get("worker_policy_snapshot")
            if not strategy or not policy:
                continue
            tokens = _plan_tokens(strategy)
            if not tokens:
                continue
            directive_tokens = _directive_token_set(policy)
            hits = sum(1 for tok in tokens if tok in directive_tokens)
            align_sum += hits / len(tokens)
            align_count += 1
        alignment = (align_sum / align_count) if align_count > 0 else 0.0

        return {
            "plan_policy_alignment": round(alignment, 4),
            "env_threat_responsiveness": round(corr, 4),
            "colony_cadence_health": round(std_dev, 2),
        }


__all__ = ["HierarchicalCoordinationPlugin"]
